core/store/src/trie/mod.rs

```diff
@@ -1361,6 +1366,21 @@ impl Trie {
         Ok(bytes.to_vec())
     }
 
+    /// Check if the column contains a value with the given `key`.
+    ///
+    /// This method is guaranteed to not inspect the value stored for this key, which would
+    /// otherwise have potential gas cost implications.
+    pub fn contains_key(&self, key: &[u8]) -> Result<bool, StorageError> {
+        let charge_gas_for_trie_node_access = self.charge_gas_for_trie_node_access;
+        if self.memtries.is_some() {
+            Ok(self.lookup_from_memory(key, charge_gas_for_trie_node_access, |_| ())?.is_some())
+        } else {
```

Comment:
Did we miss adding the lookup from flat storage? That should be preferred over trie lookup. `self.lookup_from_flat_storage(key)`

Reply:
Rest of the PR looks good! Thanks!

Reply:
Indded, I raised the same question myself in the message of the first commit. I wasn't sure if it is something that's required, superfluous or some point in between.

Reply:
IMO, it's definitely required as we would always want to first check the key in flat storage if it exists.

Reply:
Not an expert but I think we should follow the same order as in `get_optimized_ref`.

Memtrie isn't released yet but once we have that, hopefully all calls would go through that. @robin-near can give more context. While it's not on mainnet, get calls from flat_storage are preferred over trie.